ethportal-peertest/src/jsonrpc.rs

```diff
@@ -0,0 +1,158 @@
+use std::io::prelude::*;
+use std::os::unix::net::UnixStream;
+use std::slice::Iter;
+
+use hyper::{self, Body, Client, Method, Request};
+use log::info;
+use serde_json::{self, Value};
+use thiserror::Error;
+
+use trin_core::portalnet::U256;
+
+#[derive(Copy, Clone)]
+pub struct JsonRpcEndpoint {
+    pub method: &'static str,
+    pub id: &'static u8,
+}
+
+const ALL_ENDPOINTS: [JsonRpcEndpoint; 6] = [
+    JsonRpcEndpoint {
+        method: "web3_clientVersion",
+        id: &0,
+    },
+    JsonRpcEndpoint {
+        method: "discv5_nodeInfo",
+        id: &1,
+    },
+    JsonRpcEndpoint {
+        method: "discv5_routingTableInfo",
+        id: &2,
+    },
+    JsonRpcEndpoint {
+        method: "eth_blockNumber",
+        id: &3,
+    },
+    JsonRpcEndpoint {
+        method: "portalHistory_dataRadius",
+        id: &4,
+    },
+    JsonRpcEndpoint {
+        method: "portalState_dataRadius",
+        id: &5,
+    },
+];
+
+fn validate_endpoint_response(method: &str, result: &Value) {
+    match method {
+        "web3_clientVersion" => {
+            assert_eq!(result.as_str().unwrap(), "trin 0.0.1-alpha");
+        }
+        "discv5_nodeInfo" => {
+            let enr = result.get("enr").unwrap();
+            assert!(enr.is_string());
+            assert!(enr.as_str().unwrap().contains("enr:"));
+            assert!(result.get("nodeId").unwrap().is_string());
+        }
+        "discv5_routingTableInfo" => {
+            let local_key = result.get("localKey").unwrap();
+            assert!(local_key.is_string());
+            assert!(local_key.as_str().unwrap().contains("0x"));
+            assert!(result.get("buckets").unwrap().is_array());
+        }
+        "eth_blockNumber" => {
+            assert!(result.is_string());
+            assert!(result.as_str().unwrap().contains("0x"));
+        }
+        "portalHistory_dataRadius" => {
+            assert_eq!(result.as_str().unwrap(), U256::from(u64::MAX).to_string());
+        }
+        "portalState_dataRadius" => {
+            assert_eq!(result.as_str().unwrap(), U256::from(u64::MAX).to_string());
+        }
+        _ => panic!("Unsupported endpoint"),
+    };
+    info!("{:?} returned a valid response.", method);
+}
+
+impl JsonRpcEndpoint {
+    pub fn all_endpoints() -> Iter<'static, Self> {
+        ALL_ENDPOINTS.iter()
+    }
+
+    pub fn to_jsonrpc(self) -> String {
+        format!(
+            r#"
+            {{
+                "jsonrpc":"2.0",
+                "id": {},
+                "method": "{}"
+            }}"#,
+            self.id, self.method
+        )
+    }
+}
+
+pub async fn test_jsonrpc_endpoints_over_ipc() {
+    for endpoint in JsonRpcEndpoint::all_endpoints() {
+        info!("Testing over IPC: {:?}", endpoint.method);
+        let mut stream = UnixStream::connect("/tmp/trin-jsonrpc.ipc").unwrap();
```

Comment:
If we want to keep the direction to support cross-client testing in this crate, probably it is worth including the socket path as a parameter to CLI. I guess we can keep the default value as `/tmp/trin-jsonrpc.ipc`.